sys/psa_crypto: ed25519 private key {ex,im}port

mguetschow opened the pull request:
### Contribution description

This PR adds support for ed25519 private keys to `psa_import_key()` and adds the corresponding `psa_export_key()` function. For public asymmetric keys, `psa_export_key()` should behave like `psa_export_public_key()`. The backend implementations are used to derive the public key from the private key passed to `psa_import_key()`.

Note that exporting and importing private keys from and to secure elements is not yet properly handled and will likely break for now. Maybe @Einhornhool could give a hint on where and how to test whether the key is stored on a secure element, to be able to report an error to the user.

### Testing procedure

Two tests have been added to `tests/sys/psa_crypto` and tested successfully with hardware support on the nRF52840dk and using the software implementation on native.

Teufelchen1 commented on sys/psa_crypto/include/psa_crypto_algorithm_dispatch.h:
```diff
@@ -175,6 +175,21 @@ psa_status_t psa_algorithm_dispatch_verify_message( const psa_key_attributes_t *
+ * @param data_length
+ * @param slot
+ * @param bits
+ * @return psa_status_t
```
```suggestion
* @return  @ref PSA_SUCCESS
*          @ref PSA_ERROR_INVALID_ARGUMENT
 ...
```
(Maybe get's a bit long here to list all return values? Not sure if it is "better".
Just checked some established psa headers. They also do the `@return psa_status_t`. :thinking:
Feel free to dismiss this comment.)

mguetschow replied:
There are two different kinds of header files for PSA crypto right now: The "public" one that is supposed to be used by applications and other modules using the API (`psa/crypto.h` found in `sys/include/psa_crypto`) and several ones that are only used internally by the PSA crypto implementation in RIOT (found in `sys/psa_crypto/include`).

The latter ones for now contain much less thorough documentation than the former one. Would be nice to change that, but maybe not in scope of this PR?

Einhornhool replied:
It would probably be nice to have more documentation. I was not very motivated to do that, because many parameters and return values are just the same as for the public functions, but of course that was just me being lazy.

I do agree that this is not in scope for this PR, though.

mguetschow replied:
I'd keep it as is since we agreed it is out of scope for this PR.

Teufelchen1 commented on tests/sys/psa_crypto/main.c:
```diff
@@ -22,37 +22,163 @@
+/**
+ * Exporting and re-importing a private Ed25519 key should result in the same public key and signature.
+ */
+static void test_export_import_ed25519(void)
```
```suggestion
static void test_exported_key_is_identical_when_imported_again_ed25519(void)
```

Not the ideal suggestion but I want to emphasize that this test does NOT test the export and import functions. Instead, it tests integration / inter-operation of those.

If you want to test the functions in general, insert dedicated tests for that. In these test you could then go ahead and answer questions like: What happens when the key parameter is null? Do I get the correct status when the key is bigger than the buffer? etc..

(I am being pedantic here, sorry :angel:)

mguetschow replied:
I know the tests are lacking a lot in terms of coverage for PSA Crypto right now. I tried to convey the intent of this specific test with the comment above, but I agree it might be better to use a descriptive function name instead.

Teufelchen1 reviewed (changes requested):
Hey hey :mouse: 

just a short review, so this PR doesn't go stale :smiley: 
I think @Einhornhool is going to review this thoroughly?

mguetschow commented on sys/psa_crypto/psa_crypto_algorithm_dispatch.c:
```diff
@@ -427,6 +427,66 @@ psa_status_t psa_algorithm_dispatch_generate_key(   const psa_key_attributes_t *
+            }
+        }
+
+        // save private key data
```
Good point, time to internalize the [coding convention](https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#comments) a bit more.

Einhornhool reviewed (commented):
Thank you for the work! I have some questions :)

Einhornhool commented on tests/sys/psa_crypto/main.c:
```diff
@@ -22,37 +22,163 @@
 {
     EMB_UNIT_TESTFIXTURES(fixtures) {
         new_TestFixture(test_init_twice),
+        new_TestFixture(test_export_import_ed25519),
```
This function has been renamed I think

mguetschow replied:
Thanks, fixed.

mguetschow commented on sys/psa_crypto/psa_crypto_algorithm_dispatch.c:
```diff
@@ -427,6 +427,66 @@ psa_status_t psa_algorithm_dispatch_generate_key(   const psa_key_attributes_t *
+#endif
+#if IS_USED(MODULE_PSA_ASYMMETRIC_ECC_ED25519)
+        case PSA_ECC_ED25519:
+            return psa_derive_ecc_ed25519_public_key(key_data, pubkey_data, key_bytes, pubkey_data_len);
```
The PSA spec defines the standard key export type for ed25519 keypairs [to be the private key only](https://arm-software.github.io/psa-api/crypto/1.1/api/keys/management.html#key-formats), and [requires the import to support at least that format](https://arm-software.github.io/psa-api/crypto/1.1/api/keys/management.html#c.psa_export_key). I just went with only supporting that format, which means that ed25519 keypairs are always just the private keys during import. So deriving the public key is always necessary.

Einhornhool replied:
Aaah ok, thank you for explaining.

mguetschow commented on sys/psa_crypto/psa_crypto_algorithm_dispatch.c:
```diff
@@ -427,6 +427,66 @@ psa_status_t psa_algorithm_dispatch_generate_key(   const psa_key_attributes_t *
+        }
+
+        /* save private key data */
+        memcpy(key_data, data, data_length);
```
Good suggestion, changed.

mguetschow commented on sys/psa_crypto/psa_crypto.c:
```diff
@@ -1257,13 +1327,8 @@ static psa_status_t psa_builtin_export_public_key( const uint8_t *key_buffer,
     }
 
     if (data_size < key_buffer_size) {
-        DEBUG("PSA Crypto Builtin Export Key: Output buffer too small\n");
```
Hum, no idea why I removed that. Added it back now.

mguetschow commented on sys/psa_crypto/psa_crypto.c:
```diff
@@ -1220,16 +1220,86 @@ psa_status_t psa_destroy_key(psa_key_id_t key)
+        return status;
+    }
+
+    if ((data_size == 0) ||
```
Good catch, removed that now.

Einhornhool reviewed (approved):
LGTM

mguetschow commented:
> > Note that exporting and importing private keys from and to secure elements is not yet properly handled and will likely break for now. Maybe @Einhornhool could give a hint on where and how to test whether the key is stored on a secure element, to be able to report an error to the user.
> 
> This can be easily determined by the key location. There is a range of location values reserved for secure elements (either `PSA_KEY_LOCATION_PRIMARY_SECURE_ELEMENT` or a value between `PSA_KEY_LOCATION_SE_MIN` and `PSA_KEY_LOCATION_SE_MAX`). If a key has one of those locations, it is stored on an SE and can't be exported.

I think this should still be adressed before proceeding to merging. Maybe you could try it out with a secure element and report on the current behavior.

Einhornhool commented:
> > > Note that exporting and importing private keys from and to secure elements is not yet properly handled and will likely break for now. Maybe @Einhornhool could give a hint on where and how to test whether the key is stored on a secure element, to be able to report an error to the user.
> > 
> > 
> > This can be easily determined by the key location. There is a range of location values reserved for secure elements (either `PSA_KEY_LOCATION_PRIMARY_SECURE_ELEMENT` or a value between `PSA_KEY_LOCATION_SE_MIN` and `PSA_KEY_LOCATION_SE_MAX`). If a key has one of those locations, it is stored on an SE and can't be exported.
> 
> I think this should still be adressed before proceeding to merging. Maybe you could try it out with a secure element and report on the current behavior.

Damn, sorry, I forgot about this.

`psa_import_key` returns `PSA_ERROR_NOT_SUPPORTED`, because the driver does not implement this function, yet, which is the way it's supposed to be (no changes required).

`psa_export_key` returns `PSA_ERROR_INVALID_ARGUMENT`, because https://github.com/RIOT-OS/RIOT/blob/03c05f7fe1a7edee8ace4a04557d769054ff1264/sys/psa_crypto/psa_key_slot_mgmt/psa_crypto_slot_management.c#L479 returns a nullpointer, which is caught by the `psa_builtin_export_key` function. 
It would be better to check the key location before calling `psa_get_key_data_from_key_slot` and then returning `PSA_ERROR_NOT_SUPPORTED` or `PSA_ERROR_NOT_PERMITTED`.

mguetschow commented:
Thanks for testing! See [this diff](https://github.com/RIOT-OS/RIOT/pull/20334/commits/0fc4e132b8cbebdac8e76fb0a2b27948aefdb56b#diff-e56967782ddd1d30cd091927ced79e26052070cd4e2a1f3a7904d753d761e2afL1290-R1295) for the changes to fix the `psa_export_key` behavior.

On the way I noticed that I had actually broken the key import in previous fixup commits (I guess that's what the tests are for :P). Fixed that and also the `TEST_ASSERT_PSA` macro from going into an infinite loop if called within the cleanup part of the code.

May I squash?

mguetschow commented:
Wow that is a difficult one :thinking: I actually expect the other `tests/sys/psa_crypto_*` tests to fail too because of boards missing in `Makefile.ci`.

I also found a related issue along the way which is not a regression and therefore out of scope for this PR. Opened #20468 as a follow-up for it.

mguetschow commented:
> I actually expect the other `tests/sys/psa_crypto_*` tests to fail too because of boards missing in `Makefile.ci`.

They were not since they symlink to `examples/psa_crypto` instead of  `tests/sys/psa_crypto`.

@Einhornhool this should now be ready for final review, then I will squash all the fixups together.

mguetschow commented:
Thanks for the renewed review! I've just squashed and rebased onto current master.

Maybe @Teufelchen1 could have a look at the documentation discussion above and provide a proxy-ack?
